src/java/org/apache/cassandra/index/sai/disk/v1/SSTableIndexWriter.java

```diff
@@ -213,18 +221,52 @@ else if (shouldFlush(sstableRowId))
             currentBuilder = newSegmentBuilder();
         }
 
-        if (term.remaining() == 0) return;
+        if (term.remaining() == 0)
+            return;
 
-        if (!TypeUtil.isLiteral(type))
-        {
-            limiter.increment(currentBuilder.add(term, key, sstableRowId));
-        }
-        else
+        updatesInFlight.incrementAndGet();
+        ForkJoinPool.commonPool().submit(() -> {
+            try
+            {
+                addTermInternal(term, key, sstableRowId, type);
+            }
+            finally
+            {
+                updatesInFlight.decrementAndGet();
+            }
+        });
+    }
+
+    private void addTermInternal(ByteBuffer term, PrimaryKey key, long sstableRowId, AbstractType<?> type)
+    {
+        // VSTODO limiter increases asynchronously, so we may exceed the limit.  how big a problem is this?
```

Comment:
Isn't the expensive part adding to the index? If so, we could apply the limiter logic before dispatching to an executor pool.

Reply:
right, the problem is that the index is what tells us how much space got used